Add tutorial for `sct_compute_compression` in documentation (#4162)
spinalcordtoolbox/spinalcordtoolbox: documentation/source/user_section/tutorials/shape-analysis/normalizing-morphometrics-compressions/generating-necessary-inputs.rst

2. Vertebral labels
3. Compression labels (``t2_compressed_compression_labels.nii.gz``)

1. Spinal cord segmentation
----------------------------
.. code:: sh

   sct_deepseg_sc -i t2_compressed.nii.gz -c t2 -qc ~/qc_singleSubj

:Input arguments:
   - ``-i`` : Input image.
   - ``-c`` : Contrast of the input image.
   - ``-qc`` : Directory for Quality Control reporting.

:Output files/folders:
   - ``t2_compressed_seg.nii.gz`` : Spinal cord segmentation.

2. Label the spinal cord vertebral levels
------------------------------------------
.. code:: sh

   sct_label_vertebrae -i t2_compressed.nii.gz -s t2_compressed_seg.nii.gz -c t2 -qc ~/qc_singleSubj

:Input arguments:
   - ``-i`` : The input image file.
   - ``-s`` : Spinal cord segmentation.
   - ``-c`` : Contrast of the input image.
   - ``-qc`` : Directory for Quality Control reporting.

:Output files/folders:
   - ``t2_compressed_seg_labeled.nii.gz`` : Spinal cord vertebral labels.

3. Generate spinal cord compression labels
-------------------------------------------

1. Open the image in FSLeyes. If the image is not 3D, open both axial and sagittal images in FSLeyes.

.. code:: sh

   fsleyes t2_compressed.nii.gz &

> 2. Look at the clinical data provided with compression information to know at which level the compressions are located.

Reviewer: ```suggestion
2. Look at the clinical data provided with compression information to know at which level the compression(s) are located.
```

3. Locate the compression in the sagittal view.
4. In the axial view, toggle to the maximum compressed slice around the compression.
5. Click on alt+E (Windows) or option+E (macOS) to open edit mode. Select size 1 and click on the pencil.
6. Create an empty mask alt+N (Windows) or option+N (macOS).
7. Place the label at the center of the spinal cord of the axial image.
8. Repeat steps 3 to 7 for the number of compressions.
9. Save with the filename ``t2_compressed_labels-compression.nii.gz`` and quit.

You can also look at the example using an automatic script when multiples images require compression labels: 

> https://github.com/spinalcordtoolbox/manual-correction/wiki#manual-labeling-of-spinal-cord-compression

Reviewer: Maybe we could also include a screenshot (like this [one](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3984#issuecomment-1487394718)) or a short video on how to do the compression labeling.

.. _generating-necessary-inputs:

Generate the necessary input files
#############################################################

The ``sct_compute_compression`` command uses the shape metric of the spinal cord at the level of compression and computes a ratio with non-compressed levels above and below the compression site.

To use ``sct_compute_compression``, we need the 3 following input files:

1. Spinal cord segmentation
2. Vertebral labels
3. Compression labels (``t2_compressed_label-compression.nii.gz``)

> You can also get these files by downloading :sct_tutorial_data:`data_normalizing-morphometrics-compression.zip`.

Reviewer 2: Note that the current [`data_normalizing-morphometrics-compression.zip`](https://drive.google.com/file/d/1515MMcyiAqwmegpug50I5VLUrOvuU_gs/view) link is quite large, because it contains the straightening warping fields and etc. But, I think we only need a few of those files to cover the steps on this page?

I can generate a smaller `.zip` file with only the necessary files when I update the `sct_tutorial_data` dataset. :)

Author: yes that would be great, I included the wrong folder with everything oups!

1. Spinal cord segmentation
----------------------------
.. code:: sh

   sct_deepseg_sc -i t2_compressed.nii.gz -c t2 -qc ~/qc_singleSubj

:Input arguments:
   - ``-i`` : Input image.
   - ``-c`` : Contrast of the input image.
   - ``-qc`` : Directory for Quality Control reporting.

:Output files/folders:
   - ``t2_compressed_seg.nii.gz`` : Spinal cord segmentation.

2. Label the spinal cord vertebral levels
------------------------------------------
.. code:: sh

   sct_label_vertebrae -i t2_compressed.nii.gz -s t2_compressed_seg.nii.gz -c t2 -qc ~/qc_singleSubj

:Input arguments:
   - ``-i`` : The input image file.
   - ``-s`` : Spinal cord segmentation.
   - ``-c`` : Contrast of the input image.
   - ``-qc`` : Directory for Quality Control reporting.

:Output files/folders:
   - ``t2_compressed_seg_labeled.nii.gz`` : Spinal cord vertebral labels.

3. Generate spinal cord compression labels
-------------------------------------------

> 1. Open the image in FSLeyes. If the image does not have sufficient axial and sagittal resolution, open two individual images (axial image and sagittal image) in FSLeyes.

Reviewer 2: This is a very basic question, but: What does this comment mean?

I ask because, in the example file (`t2_compression.nii.gz`), I'm assuming we do have sufficient resolution, so this advice doesn't apply here.

Because I don't have a concrete example, I don't know what the alternate case ("not sufficient resolution") even looks like. I also don't know what it means to open two individual images in FSLeyes. Does the user need to have two separate images? Does the cursor sync up when they're both opened at once What if they don't have two images?

Reviewer: Excellent question! Thanks!

> I ask because, in the example file (t2_compression.nii.gz), I'm assuming we do have sufficient resolution, so this advice doesn't apply here.

Correct, `t2_compressed.nii.gz` has 0.8x0.8x0.8 mm resolution. The resolution is so-called isotropic (i.e., equivalent in all planes). It allows both 1) the computation of morphometric measures from the axial plane and 2) the identification of vertebral levels and compression level(s) in the sagittal plane. This is why we need only a single image here.

> I don't know what the alternate case ("not sufficient resolution") even looks like. I also don't know what it means to open two individual images in FSLeyes.

I admit that the term "not sufficient resolution" is not very specific here. It was supposed to mean that the image is, for example, axially acquired with a resolution of 0.5x0.5x4 mm. In such an example, 0.5x0.5 is great for the computation of morphometric measures, but 4mm slice thickness (5x higher compared to the 0.8mm image) is not ideal for vertebral levels and compression level(s). Therefore, we would open two individual images (for example, one sagittal and one axial) instead.

> Does the user need to have two separate images?

Yes. 

> Does the cursor sync up when they're both opened at once? What if they don't have two images?

Yes, FSLeyes allows you to open multiple images in a single session. The cursor is synchronized across them.

---

The approach with two images also requires their coregistration using `identity 1` during processing (see example [here](https://github.com/sct-pipeline/dcm-metric-normalization/blob/d985ac0ca24d6253bfe717310f4c6a3529243b60/scripts/process_data_dcm-zurich.sh#L157-L164)) and bringing sagittal disc labels to axial image (see [here](https://github.com/sct-pipeline/dcm-metric-normalization/issues/9)). Since this is a kind of "non-standard" or "advanced" processing, which we do not cover in other tutorials, we can maybe totally remove the sentence about two images to not confuse users:

```
1. Open the image in FSLeyes.
```

Reviewer: Done in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4162/commits/b2406ecfbbff0891c2e421948bcab42c0e2ee408

.. code:: sh

   fsleyes t2_compressed.nii.gz &

> 2. Look at the clinical data provided with compression information to know at which level the compression(s) are located.

Reviewer 2: I'm also confused about what "clinical data" means in this context, since we don't provide any example clinical data in our tutorial dataset. 

Is this something that our users will 100% have? Will this step be confusing to our users if they don't have clinical data?

Reviewer: Agree, this might be confusing. By clinical data, we mean information from clinicians about the level of compression(s), for example: 

|  participant_id | compression level  |
|---|---|
|  sub-001 | C3/C4  |
|  sub-002 | C6/C7  |

Maybe we can rephrase this:

```
2. If clinical data with information about compression level(s) is available, look at them to know at which level the compression(s) is located.
```

Reviewer: Solved in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4162/commits/144599c93f400758eb939b6c3b25282484d74d27

3. Locate the compression in the sagittal view.

.. figure:: https://raw.githubusercontent.com/spinalcordtoolbox/doc-figures/sb/4158-add-tutorial-sct-compute-compression/spinalcord-compresssion-norm/localizing_compression_sag.png

>    :align: center

Reviewer 2: Which direction should I be locating in the sagittal view? Just like, making sure to click on the correct IS slice?

Reviewer: Sorry, I do not understand. Can you please clarify what is not clear here?

Reviewer 2: Apologies for the unclear question.  What I mean is:

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/c315be97-68b7-44e8-8943-b9e7932cb0d9)

In the sagittal view, should the user be clicking on the center of the spinal cord where there is compression? Or, the anterior or posterior part of the cord? Or, does it not really matter, since the AP and LR coordinates will be fine-tuned when selecting a point in the axial view?

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/fe20ba1e-e98c-4ad9-bd82-afab79aa5179)

Reviewer 2: Ah, I see now. It doesn't matter too much because in the next step, we toggle the slice to determine the one with the greatest compression.

Reviewer 2: Maybe we could update the wording then to say something like:

```suggestion
3. Locate the compression in the sagittal view. (The location does not need to be precise, as we will fine-tune the selection in the next step.)

.. figure:: https://raw.githubusercontent.com/spinalcordtoolbox/doc-figures/sb/4158-add-tutorial-sct-compute-compression/spinalcord-compresssion-norm/localizing_compression_sag.png
   :align: center
```

Author: yes exactly! we can maybe add more details to the instructions!

> 4. In the axial view, toggle to the maximum compressed slice around the compression.

Reviewer 2: I got confused at this step, since there are no pictures.

I wasn't sure what the "maximum compressed slice" was supposed to be. Is it the LR slice, the AP slice, or the IS slice? In which direction should I be toggling?

Reviewer: > (I got a bit confused at this step, since there are no pictures.)

Do you think that a short video tutorial would be better than static screenshots?

Author: since we specify in the axial view, it is the axial slice (IS slice)

Reviewer 2: > since we specify in the axial view, it is the axial slice (IS slice)

Ahhhh okay, I get it now!! So, this means "In the axial view, increase/decrease the last "voxel location" value to view the axial slices above and below the compression site. This will allow you to determine which axial slice has the greatest compression."

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/c41a7063-9a57-46dd-abeb-8027dbd6ac37)

Reviewer 2: > Do you think that a short video tutorial would be better than static screenshots?

I don't think a video is 100% necessary? I think I just needed the wording clarified to me. :)

Author: yes!

5. Click on alt+E (Linux/Windows) or option+E (macOS) to open edit mode. Select size 1 and click on the pencil.

.. figure:: https://raw.githubusercontent.com/spinalcordtoolbox/doc-figures/sb/4158-add-tutorial-sct-compute-compression/spinalcord-compresssion-norm/edit_mode.png
   :align: center

6. Create an empty mask alt+N (Linux/Windows) or option+N (macOS).

.. figure:: https://raw.githubusercontent.com/spinalcordtoolbox/doc-figures/sb/4158-add-tutorial-sct-compute-compression/spinalcord-compresssion-norm/create_mask.png
   :align: center

7. Place the label at the center of the spinal cord of the axial image.

.. figure:: https://raw.githubusercontent.com/spinalcordtoolbox/doc-figures/sb/4158-add-tutorial-sct-compute-compression/spinalcord-compresssion-norm/labeling_compression.png
   :align: center

> 8. Repeat steps 3 to 7 for the number of compressions.

Reviewer 2: This step was a bit confusing to me, because I wasn't sure how many compression labels should be placed.

On the next page, I realized that there are 4 compression sites in total. So, it might be useful to provide the exact points that you used? Something like:

"Note: Our manual labeling team identified 4 compression sites, with labels placed at the following coordinates. Please place labels at those locations if you would like to recreate the sample metric values for this tutorial."

Reviewer: >  I wasn't sure how many compression labels should be placed.

This is information which should be provided by clinicians; see [this comment](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4162#discussion_r1301538363).

Reviewer 2: Maybe what we could do is update [Step 2](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4162#discussion_r1301538363) to say "In this example, we were provided with the following clinical observations: [show table with vertebrae]. So, we will be adding 4 compression labels in total."

Then, in Step 8, we can say:

```suggestion
8. Repeat steps 3 to 7 for the remaining three compression sites.
```

That way, it is clear from the outset how many compression labels the user should add during the tutorial.

> 9. Save with the filename ``t2_compressed_label-compression.nii.gz`` and quit.

Reviewer 2: ```suggestion
9. Save with the filename ``t2_compressed_labels-compression.nii.gz`` and quit.
```

The commands on the next page mention "labels" rather than "label".

> If you need to label multiple patients, you can use the `manual_correction.py` script from the [`manual-correction` repository](https://github.com/spinalcordtoolbox/manual-correction); see the example [here](https://github.com/spinalcordtoolbox/manual-correction/wiki#manual-labeling-of-spinal-cord-compression).

Reviewer 2: These are markdown links. I believe this is the equivalent RST syntax:

```suggestion
If you need to label multiple patients, you can use the ``manual_correction.py`` script from the `manual-correction repository <https://github.com/spinalcordtoolbox/manual-correction>`_; see the example `here <https://github.com/spinalcordtoolbox/manual-correction/wiki#manual-labeling-of-spinal-cord-compression>`_.
```

.. _generating-necessary-inputs:

Generate the necessary input files
#############################################################

> The ``sct_compute_compression`` command uses the shape metric of the spinal cord a the level of compression and computes a ratio with healthy levels above and below the compression site.

Reviewer: ```suggestion
The ``sct_compute_compression`` command uses the shape metric of the spinal cord at the level of compression and computes a ratio with non-compressed levels above and below the compression site.
```

To use ``sct_compute_compression``, we need the 3 following input files:

1. Spinal cord segmentation
2. Vertebral labels

> 3. Compression labels (``t2_compressed_compression_labels.nii.gz``)

Reviewer: Again, we should use the `label-compression.nii.gz` suffix to be in line with the [intranet](https://intranet.neuro.polymtl.ca/data/dataset-curation.html#derivatives-structure).

You can also get these files by downloading :sct_tutorial_data:`data_normalizing-morphometrics-compression.zip`.

1. Spinal cord segmentation
----------------------------
.. code:: sh

   sct_deepseg_sc -i t2_compressed.nii.gz -c t2 -qc ~/qc_singleSubj

:Input arguments:
   - ``-i`` : Input image.
   - ``-c`` : Contrast of the input image.
   - ``-qc`` : Directory for Quality Control reporting.

:Output files/folders:
   - ``t2_compressed_seg.nii.gz`` : Spinal cord segmentation.

2. Label the spinal cord vertebral levels
------------------------------------------
.. code:: sh

   sct_label_vertebrae -i t2_compressed.nii.gz -s t2_compressed_seg.nii.gz -c t2 -qc ~/qc_singleSubj

:Input arguments:
   - ``-i`` : The input image file.
   - ``-s`` : Spinal cord segmentation.
   - ``-c`` : Contrast of the input image.
   - ``-qc`` : Directory for Quality Control reporting.

:Output files/folders:
   - ``t2_compressed_seg_labeled.nii.gz`` : Spinal cord vertebral labels.

3. Generate spinal cord compression labels
-------------------------------------------

> 1. Open the image in FSLeyes. If the image is not 3D, open both axial and sagittal images in FSLeyes.

Reviewer: ```suggestion
1. Open the image in FSLeyes. If the image does not have sufficient axial and sagittal resolution, open two individual images (axial image and sagittal image) in FSLeyes.
```

.. code:: sh

   fsleyes t2_compressed.nii.gz &

2. Look at the clinical data provided with compression information to know at which level the compression(s) are located.
3. Locate the compression in the sagittal view.

.. figure:: https://raw.githubusercontent.com/spinalcordtoolbox/doc-figures/sb/4158-add-tutorial-sct-compute-compression/spinalcord-compresssion-norm/localizing_compression_sag.png
   :align: center

4. In the axial view, toggle to the maximum compressed slice around the compression.
5. Click on alt+E (Windows) or option+E (macOS) to open edit mode. Select size 1 and click on the pencil.

.. figure:: https://raw.githubusercontent.com/spinalcordtoolbox/doc-figures/sb/4158-add-tutorial-sct-compute-compression/spinalcord-compresssion-norm/edit_mode.png
   :align: center

> 6. Create an empty mask alt+N (Windows) or option+N (macOS).

Reviewer: ```suggestion
6. Create an empty mask alt+N (Linux/Windows) or option+N (macOS).
```

.. figure:: https://raw.githubusercontent.com/spinalcordtoolbox/doc-figures/sb/4158-add-tutorial-sct-compute-compression/spinalcord-compresssion-norm/create_mask.png
   :align: center

7. Place the label at the center of the spinal cord of the axial image.

.. figure:: https://raw.githubusercontent.com/spinalcordtoolbox/doc-figures/sb/4158-add-tutorial-sct-compute-compression/spinalcord-compresssion-norm/labeling_compression.png
   :align: center

8. Repeat steps 3 to 7 for the number of compressions.

> 9. Save with the filename ``t2_compressed_labels-compression.nii.gz`` and quit.

Reviewer: Fix suffix to `label-compression.nii.gz`  to be in line with the [intranet](https://intranet.neuro.polymtl.ca/data/dataset-curation.html#derivatives-structure).

> You can also look at the example using an automatic script when multiples images require compression labels: 
> https://github.com/spinalcordtoolbox/manual-correction/wiki#manual-labeling-of-spinal-cord-compression

Reviewer: ```suggestion
If you need to label multiple patients, you can use the `manual_correction.py` script from the [`manual-correction` repository](https://github.com/spinalcordtoolbox/manual-correction); see the example [here](https://github.com/spinalcordtoolbox/manual-correction/wiki#manual-labeling-of-spinal-cord-compression).
```
